### ISSUE TITLE
Court User: Ability to put a “hold” on a case

### ISSUE THREAD

**Issue by mariahkannenberg** [flexion/ef-cms#1188]
As a Court User, in order to prevent a case from being scheduled for trial, I need the ability to put a "hold" on a case.

There are certain circumstances where the Court may not want a specific case to be calendared for trial.  This might happen when there is a specific, new issue that needs a test case to be decided first, or if the Court is waiting for all the cases in a group to be filed so that they can get calendared together.  In addition, while a case is waiting to be scheduled for trial, a party might file something that needs to be addressed before the case can go to trial.  As a result, Court Users need to be able to put a "hold" on a case at any time to prevent it from being scheduled for trial.

## Pre-Conditions
* Case exists in the system

## Acceptance Criteria
* Any Court User can put a "hold" on a case at any time, which would prevent that case from being scheduled for a trial session until that "hold" is removed
* Court Users could at any time view a list of cases per trial location that are in "hold" status
* List/report that is generated should list docket number and case name
* When scheduling a trial session, cases with a "hold" status would not get scheduled, even if they were "next in line" sequentially
* A Court User could at any time remove the "hold" status for a case, thereby freeing it up to be scheduled

## Notes
* Consider a case status of `General Docket - At Issue (Hold for Trial)` (Confirmed w/Jessica - status not best way to proceed.)
* From meeting with Tina and Rhonda (10/8/19) - only cases that have the "General Docket (At Issue) - Ready for Trial" need to have a "Block" option. The "block" is not a case status, but should be a flag that can accompany the status.

## Security Considerations
* Does this work make you nervous about privacy or security?
* Does this work make major changes to the system?
* Does this work implement new authentication or security controls?
* Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Tasks

- [x] 1. add the "block from trial" button onto the case information - trail information card (computed for hiding / showing) (Cody #3294)
- [x] 2. block from trial modal (sequence for opening) (Cody #3294)
- [x] 3. create the block case sequence which should hit a block case endpoint via proxy, show green success alert after blocking (Cody #3294)
- [x] 4. on the case information tab, show the blocked for trial red alert and button (computeds) (Cody #3294)
- [x] 5. work on the remove block from case modal, sequence for opening it (Cody #3294)
- [x] 6. remove block sequence which uses the proxy to hit the unblock endpoint and green success alert (Cody #3294)
- [x] 7. block endpoint (interactor, serverless.yml, lambda, persistence) (Cody #3289)
- [x] 8. unblock endpoint (interactor, serverless.yml, lambda, persistence) (Cody #3289) 
- [x] 9. add flag / methods onto case entity for setting it as blocked (Rachael #3285)
- [x] 10. need to add a blocked cases link into the reports dropdown of the top navigaton (Rachael #3296)
- [x] 11. create blocked cases page route (Rachael #3296)
- [x] 12. create blocked cases page table, computed for formatting (Rachael #3296)
- [x] 13. create blocked cases page select criteria, sequence on change (Rachael #3296)
- [x] 14. create a "get blocked cases" endpoint (interactor, serverless.yml, lambda, persistence) (Rachael #3293)
- [x] 15. create the blocked indicator on the case detail header with the hand-paper icon (Cody)
- [x] 16. Filter out blocked cases when setting for trial (Rachael, Cody #3301)
- [x] 17. Elasticsearch stress testing
- [x] 18. Cleanup
  - [x] clear form when going to blocked cases report (Rachael #3301)
  - [x] clean up UI css/styles
- [x] 19. Look into elasticsearch error on case 102-19 on dev
- [x] 20. Add Date Blocked (Rachael, Cody #3304)
  - [x] to alert-error on Case Detail (Rachael, Cody #3304)
  - [x] to Blocked Cases Report table (Rachael, Cody #3304)
- [x] 21. Add Empty state message to Blocked Cases Report (Kristen #3305)
- [x] 22. Sort blocked cases in table by docket number (Rachael #3309)

## Definition of Done (Updated 8-28-19)
**Product Owner**
 - [x]  Acceptance criteria have been met

**UX**
 - [x] Business test scenarios to meet all acceptance criteria have been written
 - [x] Usability has been validated
 - [x] Wiki has been updated (if applicable) 
- [x] Story has been tested on a mobile device (for external users only)

**Engineering**
- [x] Automated test scripts have been written
- [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Module dependencies are up-to-date and are at the latest resolvable version (npm update)
 - [x] Errors in Sonarcloud are fixed https://sonarcloud.io/organizations/flexion-github/projects
 - [x] Lambdas include CloudWatch logging of users, inputs and outputs
 - [x] Code refactored for clarity and to remove any known technical debt 
 - [x] Deployed to the dev environment
- [ ] Deployed to the stage environment

**Comment by klohman** [comment]
### Design Updates 

#### Case Detail 
![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/793bea9c-393f-4343-a0eb-35736a837381)

* Add "Block from Trial" button to Trial Information card 
* On click of button, open confirmation modal 

#### Confirmation Modal 
![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/c40ffee0-110e-4040-a4dc-558c3d1a1b6d)

On click of "No, cancel", modal closes and no changes to case status 

On click of "Yes, block from trial" : 

* Modal closes 
* Case status is changed to "General Docket (At Issue) - Block from Trial" 
   - **This is a new case status**
* Case status label background is changed to #b51d09 - secondary dark 
* "Block from Trial" button is hidden 
* "Remove Block" destructive link is displayed 
* Case will not be included when trial session is "Set" 

![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/ffbf11ff-a110-4f2e-951c-288159cad720)

#### Trial Session Detail 
![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/eec0dfba-d679-45f5-95c8-a27fe0bbc194)

* Update label on session detail - "[TERM]: UPCOMING - NOT SET" 
   - "Not Set" should display on label until the trial session is set (aka calendared) 
* Add new "Blocked Cases" tab 

**Blocked Cases tab**
* Display only when the trial session has not been set   
* Include all cases with the following criteria 
  - Case's request trial location equals the session location 
  - Case type = session's case type (aka Regular, Small, etc.) 
  - Case status = General Docket (At Issue) - Blocked from Trial 
* Display cases in docket number order

**Comment by klohman** [comment]
### 1188 - Test Scenarios 

#### Scenario 1) User puts a hold (block) on a case that is “General Docket (At Issue) - Ready for Trial”
* Create case
    * Trial Location - Boston, MA
    * Regular case procedure 
    * Case status “General Docket (At Issue) - Ready for Trial”
* Login as docket clerk 
* Navigate to case xxx-xx
* Click “Block From Trial” and confirm 
Expected Results 
- [ ] On click of “Yes, block from trial”, modal closes
- [ ] Case detail page displays success message
- [ ] Case status label changes to “General Docket (At Issue)- Ready for Trial - Blocked” with red background 
- [ ] “Block from Trial” button is hidden
- [ ] “Remove Block” link is displayed 

#### Scenario 2) Users cannot put block on case that is not in General Docket (At Issue) Ready for Trial status 
* Create case
    * Trial Location - Boston, MA
    * Regular case procedure 
    * Case status “General Docket (Not at Issue)”
* Login as docket clerk 
* Navigate to case xxx-xx
Expected Results 
- [ ] “Block from Trial” button is hidden 

#### Scenario 3) User views blocked cases for trial session 
* Create trial session for Boston, MA Regular cases 
* Create 2 cases 
    * Case 1 - trial location Boston, MA, Regular case procedure- Status- Ready for Trial 
    * Case 2 - trial location Boston, MA, Regular case procedure- Status- Ready for Trial
* [Scenario 1] 
* Navigate to Trial Session for Boston, MA 
* Click on Blocked Cases 
Expected results 
- [ ] Cases 1 and 2 display under Eligible Cases for Boston, MA session 
- [ ] Case xxx-xx displays under Blocked Cases for Boston, MA session 

#### Scenario 3) User sets calendar for trial session with blocked cases 
* [Scenario 3] 
* Click “Set Calendar” 
Expected results 
- [ ] Case 1 and Case 2 status changes to “Calendared” 
- [ ] Case 1 and Case 2 display in Open Cases tab for Boston, MA session 
- [ ] Case xxx-xx does not display in Open Cases for Boston, MA session 
- [ ] Case xxx-xx status remains “General Docket (At Issue)- Blocked from Trial” 

#### Scenario 4) User removes block from a case 
* Create new trial session for Boston, MA Regular cases 
* Navigate to case xxx-xx (from Scenario 1) 
* Remove block from case 
* Navigate to new Boston, MA 
Expected results 
On click of “remove block”
- [ ] modal closes and Case Detail displays success message
- [ ] Case status label updates to "General Docket (At Issue) - Ready for Trial and background color is updated to white
- [ ] “Blocked" flag is removed from case
- [ ] “Block From Trial" button is displayed in Trial Information card
- [ ] Case is displayed under Eligible Cases for new Boston, MA session

**Comment by klohman** [comment]
### Blocked Cases 

#### Update to Navigation  
* Add "Blocked Cases" link to "Reports" submenu 
* On select of "Blocked Cases", open Blocked Cases screen 

![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/dfc1f925-14b7-4a7e-a2e0-e82139795bbe)

#### Blocked Cases report

**Default view**
* no trial location selected 
* message displays "Select a trial location to view blocked cases"
![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/2430dc90-8c58-4bf3-9dc6-a432af692214)


**Trial Location**
* dropdown should contain the full list of trial locations 
* on select of a trial location from dropdown, display table 

![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/34726bb4-f849-4451-afbd-f5a9f60cc5d1)

**Blocked Cases table**

* Display selected trial location as table section header (example in design - Birmingham, Alabama
* Display count of number of blocked cases for selected trial location 

* Display all cases with the following criteria: 
- selected trial location = place of trial on case 
- case has a "blocked from trial" flag 

* Reason column = displays text entered in "Why are you blocking this case" field 
* Cases display in docket number order (oldest to newest)

**Comment by klohman** [comment]
### 1188 - Test Scenarios 

#### Scenario 1) User puts a hold (block) on a case that is “General Docket (At Issue) - Ready for Trial”
* Create case 1
    * Trial Location - Boston, MA
    * Regular case procedure 
    * Case status “General Docket (At Issue) - Ready for Trial”
* Login as docket clerk 
* Navigate to case xxx-xx
* Click “Block From Trial”, enter reason and save  
Expected Results 
- [ ] On click of “Yes, block from trial”, modal closes
- [ ] Case detail page displays success message
- [ ] “Blocked” label displays in docket header 
- [ ] “Block from Trial” button is hidden
- [ ] “Block” warning message is displayed 

#### Scenario 2) Users puts block on case that is not in General Docket (At Issue) Ready for Trial status 
* Create case 2
    * Trial Location - Boston, MA
    * Regular case procedure 
    * Case status “General Docket (Not at Issue)”
* Login as docket clerk 
* Navigate to case xxx-xx
Expected Results 
- [ ] On click of “Yes, block from trial”, modal closes
- [ ] Case detail page displays success message
- [ ] “Blocked” label displays in docket header 
- [ ] “Block from Trial” button is hidden
- [ ] “Block” warning message is displayed 

#### Scenario 3) User does not see blocked cases in eligible case list for trial session
* Create case 3 
    * Trial Location = Boston, MA 
    * Regular case procedure 
    * Status = General Docket (At Issue)- Ready for Trial 
* Create trial session for Boston, MA, Regular Case procedure 
* Navigate to new Boston, MA session 
Expected results 
- [ ] Case 3 displays in Eligible Cases 
- [ ] Case 1 and Case 2 do not display under eligible cases 

#### Scenario 4) User sets calendar for trial session with blocked cases 
* [Scenario 3]
* Click on “Set Calendar”
Expected results 
- [ ] Case 3 status changes to “Calendared” 
- [ ] Case 1 and Case 2 status does not change 
- [ ] Case 3 displays in Open Cases for Boston, MA session 


#### Scenario 5) User views all blocked cases for a trial location 
* Navigate to Reports > Blocked Cases 
* Select Boston, MA from dropdown 
Expected Results 
- [ ] Case 1 and case 2 display in table  


#### Scenario 6) User removes block from a case 
* Create new trial session for Boston, MA, regular case procedure 
* Navigate to Case 2
* Click on Case Info tab 
* Click on Remove block and confirm
Expected results 
On click of “remove block”
- [ ] modal closes and Case Detail displays success message
- [ ] “Blocked" label is removed from docket header
- [ ] “Block From Trial" button is displayed in Trial Information card
- [ ] Case is displayed under Eligible Cases for new Boston, MA session 

#### Scenario 7) User cannot block a case in “Calendared” status 
* [Scenario 4] 
* Navigate to case 3 
* View Trial Info card 
Expected results 
- [ ] “Block from Trial” button is not displayed

**Comment by JessicaMarine** [comment]
PO review 10/18/19:
* When a case is in "hold" status, it should *not* appear on the eligible cases list on a new/upcoming trial session
